10_rest/willshen8/pkg/puppy/rest.go

```diff
@@ -0,0 +1,107 @@
+package puppy
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/go-chi/chi"
+	"github.com/go-chi/render"
+)
+
+// RestHandler provides a puppy storer to store the puppies
+type RestHandler struct {
+	// sync.Mutex
+	store Storer
+}
+
+// NewRestHandler is a factory method that makes new storer
+func NewRestHandler(storer Storer) *RestHandler {
+	return &RestHandler{
+		store: storer,
+	}
+}
+
+// HandleGet gets the puppy by id and displays the results.
+func (rh *RestHandler) HandleGet(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		render.JSON(w, r, &Error{
+			Message: InvalidInput,
+			Code:    HTTPBadRequest,
+		})
+	} else {
+		puppy, readErr := rh.store.ReadPuppy(uint32(id))
+		if readErr != nil {
+			w.WriteHeader(http.StatusBadRequest)
```

Comment:
Here you're returning header response `400` bad request for proper request. It should be `404` in my mind.
The request was ok, but `resource not found` - 404

Maybe you can reuse `Error` struct. You've `Code` field there.
I can suggest 3 options with 1st preferred.

1. You can use error codes. You can replace your error codes `100x` with `40x` and then `w.WrtieHeader(err.Code)`

2. You can provide a translation layer to return `404` for `1002` and `400` for `1001 and `1001`

3. Your can use custom codes. In example `1002` have no conflicts with well-known codes, so It should be ok to just return this.

```
 n0npax@narrenturm  /tmp/go-course/10_rest/willshen8   lab10 ●   curl localhost:8888/api/puppy/33 -v
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8888 (#0)
> GET /api/puppy/33 HTTP/1.1
> Host: localhost:8888
> User-Agent: curl/7.58.0
> Accept: */*
>
< HTTP/1.1 400 Bad Request <------- HERE request is OK, just no resource found
< Date: Sun, 18 Aug 2019 06:48:46 GMT
< Content-Length: 76
< Content-Type: text/plain; charset=utf-8
<
{"message":"Puppy ID can not be found, read operation failed.","code":1002}
* Connection #0 to host localhost left intact

```

Reply:
to be honest for given request `curl localhost:8888/api/puppy/33 -v` response code should be `404`, but for
 `curl localhost:8888/api/puppy/foo -v` it should be `400`

Reply:
fixed!